### PR TITLE
[FLINK-2310] Add an Adamic Adar Similarity example

### PR DESCRIPTION
**Pull request by shghatge**
Added an Adamic Adar Similarity example to Gelly.
Added one test accordingly for the same.

### CODE REVIEW

**Comment by shghatge** [comment]
Updated PR.

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import org.apache.flink.graph.Triplet;
+import org.apache.flink.graph.example.utils.AdamicAdarSimilarityMeasureData;
+import java.util.HashSet;
+
+/**
+ * Given a directed, unweighted graph, return a weighted graph where the edge values are equal
+ * to the Adamic Acard similarity coefficient which is given as
+ * Summation of weights of common neighbors of the source and destination vertex
+ *The weights are given as 1/log(nK) nK is the degree  or the vertex
+ * See <a href="http://social.cs.uiuc.edu/class/cs591kgk/friendsadamic.pdf">Friends and neighbors on the Web</a>
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java]: @see, not just see :)

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+		public Tuple2<Double, HashSet<Tuple2<Long, Double>>>
+		reduceNeighbors(Tuple2<Double, HashSet<Tuple2<Long, Double>>> first,
+						Tuple2<Double, HashSet<Tuple2<Long, Double>>> second) {
+			first.f1.addAll(second.f1);
+			return first;
+		}
+	}
+
+	//mapfunction to initialise Vertex values from edge dataset
+	private static class mapVertices implements MapFunction<Long, Tuple2<Double, HashSet<Tuple2<Long, Double>>>> {
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java]: class name that starts with "m"? :)

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			for(Tuple2<Long, Double> t : value.f0.f1) {
+				t = new Tuple2<Long, Double> (t.f0, value.f0.f0);
+				value.f0.f1.clear();
+				value.f0.f1.add(t);
+			}
+			return value.f0;
+		}
+	}
+
+	//mapfunction to update values of vertices in the Graph
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java]: Let's be consistent and out them all with /** ...*/ when they're in front of classes

**Comment by andralungu** [comment]
Code-style freak says everything else is fine

**Comment by shghatge** [comment]
Updated PR.

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.ReduceNeighborsFunction;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.Triplet;
+import org.apache.flink.graph.example.utils.AdamicAdarSimilarityMeasureData;
+import java.util.HashSet;
+
+/**
+ * Given a directed, unweighted graph, return a weighted graph where the edge values are equal
+ * to the Adamic Acard similarity coefficient which is given as
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java]: Acard -> Adar

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+ * Summation of weights of common neighbors of the source and destination vertex
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/AdamicAdarSimilarityMeasure.java]: we usually use the term "weight" for an edge and "value" for a vertex

**Comment by vasia** [comment]
Hi @shghatge,
thank you for working on this! It's not a trivial problem :)

I have two general comments:

- I would prefer to see this as library method instead of an example. As an example, it shows how to use neighborhood methods, joinWithVertices and getTriplets, but we already have the JaccardSimilarity for all of these.

- The algorithm is actually very hard to scale. It essentially requires common neighborhood calculation. This is a real pain for large graphs with skewed vertices. @andralungu can explain to you more, since she's been working on optimizing similar problems for her work.
An additional problem that this particular implementation has is that it also stores and sends vertex values, not only vertex IDs. If you try to run this with a big graph, it will probably crush or run forever.

There are 2 things we could do:

1). Only send a list of vertex IDs and let the receiving vertex generate partial sums for each common neighbor. At the end, we can aggregate the sums per edge. For example, consider the edges (1, 2), (1, 3), (2, 3), (1, 4) and (2, 4). In the neighborhood method, vertex 3 will find that it has a common neighbors with 1, vertex 2. Then, it will emit a tuple `<1, 2, 1/log(d3)>`. In the same way, vertex 4 will find that it has a common neighbor with 1, vertex 2 and it will emit a tuple `<1, 2, 1/log(d4)>`.
The output of the neighborhood method will be a dataset of tuples that represent edges with partial sums. You can then compute the Adamic-Adar similarity by summing up the values per vertex.

2). The above solution will be OK for regular graphs, but will still have problems in the presence of skew. Another approach would be to offer an approximate Adamic-Adar similarity, by representing neighborhoods as bloom filters. In this case, the size of common neighborhoods will be over-estimated, but the computation will be faster.

@shghatge, @andralungu, let me know what you think and how you would like to proceed!

**Comment by shghatge** [comment]
Hello @vasia 
I would like to work on both versions of Adamic Adar. As the JIRA did not ask for an approximate version, it was suggested that I create another JIRA issue which will provide a library method for Adamic Adar which gives approximate solution with the use of bloom filters.
I have a query about the bloom filters. Since bloom filters only tell us whether an element belongs to the set or not, if both the vertices have Bloom filters as value, how will we know what to search for in the other set? For example. for Example for Vertex 3 '1,4,13' are set and for Vertex 5 '2,4,13' are set. Now when we use the method suggested by you, we will find that 4 and 13 are set for 5 too. Now what tuple should it emit? Do you suggest that we keep another hashtable that keeps track of a value->vertex relation? Or do we just emit 5,4,1/log(d3) and keep the hashtable as an identity map function? That would mean each vertex has n number of bits as value , where n is the number of vertices in the graph. I hope I was clear in my query. TL;DR We will have to use an identity hash function which implies that each vertex will need n bits of memory as value. Is it okay to use this much memory? If there is some other approach then please let me know. Bloom filters seem to be more useful in finding size of the intersection or union but here we need to know which Vertices are common. The only other way that I can roughly imagine is that we get the hashed edges in a dataset, just like 5,4,1/(logd3)... Use the same hash function on all the graph edges. Then Join the datasets obtained over field 1 and 2. 
Please tell me if there is any other efficient way or which one of these two you would prefer?

**Comment by shghatge** [comment]
Hello @vasia 
I would like to work on both versions of Adamic Adar. As the JIRA did not ask for an approximate version, it was suggested that I create another JIRA issue which will provide a library method for Adamic Adar which gives approximate solution with the use of bloom filters.
I have a query about the bloom filters. Since bloom filters only tell us whether an element belongs to the set or not, if both the vertices have Bloom filters as value, how will we know what to emit? For Example for Vertex 3 '1,4,13' are set and for Vertex 5 '2,4,13' are set. Now when we use the method suggested by you using logical AND we find out the intersection of the Bloom Filters. After this do you suggest that we keep another hashtable that keeps track of a value->vertex relation? Or do we just emit 5,4,1/log(d3) and keep the hashtable as an identity map function? That would mean each vertex has n number of bits as value , where n is the number of vertices in the graph. I hope I was clear in my query. TL;DR We will have to use an identity hash function which implies that each vertex will need n bits of memory as value. Is it okay to use this much memory? If there is some other approach then please let me know. Bloom filters seem to be more useful in finding size of the intersection or union but here we need to know which Vertices are common. The only other way that I can roughly imagine is that we get the hashed edges in a dataset, just like 5,4,1/(logd3)... Use the same hash function on all the graph edges. Then Join the datasets obtained over field 1 and 2. 
Please tell me if there is any other efficient way or which one of these two you would prefer?

**Comment by vasia** [comment]
Hi @shghatge!

I agree, let's deal with the approximate version as a separate issue. In the end though, it would be nice to have a single library method and an input parameter to decide whether the computation should be exact or approximate.

Regarding the bloom filter, the idea is for each vertex to build a bloom filter with its neighbors and "send" it to its neighbors. Then, each vertex can compare its own neighborhood (the exact one) with the received bloom filter neighborhoods. Take a look at how approximate Jaccard is computed in the okapi library [here](https://github.com/grafos-ml/okapi/blob/master/src/main/java/ml/grafos/okapi/graphs/similarity/Jaccard.java) (class `JaccardApproximation `).

Let me know if you have more questions :)

**Comment by shghatge** [comment]
Hello @vasia 
Thanks for the clarification. I was only thinking of changing the data structure :) But this method seems much more sensible. In any case... Should I close this pull request? Or should I commit the changes made to it according to your first suggestion?

**Comment by vasia** [comment]
You can use this PR for the exact computation, no need to open a new one!

**Comment by shghatge** [comment]
Updated PR.
I have used the first improvement suggested by @vasia to calculate exact values of Adamic Adar edges.

Since in Adamic Adar similarity algorithm vertices are given weights according to the number of their neighbors, the word 'weights' is used sometimes. Whenever it is used it means 1/log(kn) where kn is the degree of the vertex.

**Comment by shghatge** [comment]
@andralungu @vasia PR has been updated to make the code more efficient.

**Comment by andralungu** [comment]
Hi @shghatge ,

Did you also run the two examples on the cluster to make sure that the approximate version is faster?
Then you could also add some numbers to the two PRs.

**Comment by vasia** [comment]
Hi @shghatge,

let me try to explain the implementation in detail here :)

First of all, can you change this to a library method instead of an example? The modification should be very easy. You only need to move the class in the `org.apache.flink.graph.library` package and implement the `GraphAlgorithm` interface. This way, users will be able to use this method by simply calling `graph.run(new AdamicAdarSimilarity())`.

Making this a library method also means that we don't have to use only Gelly methods, i.e. we can do a few things more efficiently.

For example, in the beginning of the algorithm, you need to compute (1) the vertex "weights" and (2) the neighbor IDs for each vertex. Both these computations can be done with a single GroupReduce on the edges dataset, i.e. `edges.flatMap(...).groupBy(0).reduceGroup(...)`, where in the flatMap you simply create the opposite direction edges and in the reduceGroup you compute the neighborhood sets and degrees (size of the set) - weights.
The result will be a dataset where each vertex has a `Tuple2` value with its "weight" as the first field and its neighbors as the second.

Similarly, instead of using `getTriplets()` (which is convenient but quite expensive), you can compute the partial edge values with a single `groupReduceOnNeighbors`. 
Say you have vertices `(1, {d1, (2, 3, 4)})`, `(2, {d2, (1, 3)})`, `(3, {d3, (2, 4)})` and `(4, {d4, (1, 3)})`. 
In `groupReduceOnNeighbors`, vertex 1 will compute the following:
- neighbor 2: common neighbor 3 -> emit `(2, 3, d1)`
- neighbor 3: common neighbor 4 -> emit `(3, 4, d1)`

Finally, you can `groupBy(0, 1)` this result dataset and compute the sums to get the final adamic-adar similarities.

Let me know if this makes sense and whether you need more clarifications!

**Comment by shghatge** [comment]
I made some of the changes mentioned. But there was a problem in replacing the getTriplets() and flatMap. When I use this method
private static final class ReduceN implements NeighborsFunctionWithVertexValue<Long,Tuple2<Double, HashSet<Long>> ,Double,
									Tuple3<Long, Long, Double>> {

		@Override
		public void iterateNeighbors(
				Vertex<Long, Tuple2<Double, HashSet<Long>>> vertex,
				Iterable<Tuple2<Edge<Long, Double>, Vertex<Long, Tuple2<Double, HashSet<Long>>>>> neighbors,
				Collector<Tuple3<Long, Long, Double>> out) throws Exception {
			for(Tuple2<Edge<Long, Double>, Vertex<Long, Tuple2<Double, HashSet<Long>>>> neighbor : neighbors) {
				HashSet<Long> intersection = vertex.getValue().f1;
				intersection.retainAll(neighbor.f1.getValue().f1);
				for(long l : intersection) {
					System.out.println(neighbor.f1.getId()+" "+l);
						out.collect(new Tuple3<Long, Long, Double>(neighbor.f1.getId(), l, vertex.getValue().f0));
					}
			}
		
		}
		
	}

For some reason the

**Comment by andralungu** [comment]
Hi @shghatge,

Remember the git rebase command we talked about when you were in Berlin? You should revert the last commit and rebase instead... right now it looks like 15 other people contributed to your code ;) This is because you merged the master into your branch.

**Comment by vasia** [comment]
Any progress on this and #923 @shghatge?

**Comment by shghatge** [comment]
@vasia This is giving merge conflicts when rebasing. Since the conflicts are in commits other than mine, I am thinking of creating a new PR. I have done the similar changes in #923. If the last commit in this is okay, I will update that PR as well?

**Comment by andralungu** [comment]
I think it's safe to open a fresh PR... fixing this one would be overkill. You can also update and rebase #923 so that we can review them.

**Comment by vasia** [comment]
Hi @shghatge,
are you still working on this and #923?
If not, could you please close these 2 PRs? Thanks a lot!

**Pull request closed without merging**